include/alpaka/queue/QueueGenericNonBlocking.hpp

```diff
@@ -0,0 +1,244 @@
+            {
+                //-----------------------------------------------------------------------------
+                ALPAKA_FN_HOST static auto getDev(
+                    queue::QueueGenericNonBlocking<TDev> const & queue)
+                -> TDev
+                {
+                    return queue.m_spQueueImpl->m_dev;
+                }
+            };
+        }
+    }
+    namespace event
+    {
+        namespace traits
+        {
+            //#############################################################################
+            //! The CPU non-blocking device queue event type trait specialization.
+            template<
+                typename TDev>
+            struct EventType<
+                queue::QueueGenericNonBlocking<TDev>>
+            {
+                using type = event::EventGeneric<TDev>;
+            };
+        }
+    }
+    namespace queue
+    {
+        namespace traits
+        {
+            //#############################################################################
+            //! The CPU non-blocking device queue enqueue trait specialization.
+            //! This default implementation for all tasks directly invokes the function call operator of the task.
+            template<
+                typename TDev,
+                typename TTask>
+            struct Enqueue<
+                queue::QueueGenericNonBlocking<TDev>,
+                TTask>
+            {
+                //-----------------------------------------------------------------------------
+                ALPAKA_FN_HOST static auto enqueue(
+#if !(BOOST_COMP_CLANG_CUDA && BOOST_ARCH_PTX)
+                    queue::QueueGenericNonBlocking<TDev> & queue,
+                    TTask const & task)
+#else
+                    queue::QueueGenericNonBlocking<TDev> &,
+                    TTask const &)
+#endif
+                -> void
+                {
+// Workaround: Clang can not support this when natively compiling device code. See ConcurrentExecPool.hpp.
+#if !(BOOST_COMP_CLANG_CUDA && BOOST_ARCH_PTX)
+                    queue.m_spQueueImpl->m_workerThread.enqueueTask(
+                        task);
+#endif
+                }
+            };
+            //#############################################################################
+            //! The CPU non-blocking device queue test trait specialization.
+            template<
+                typename TDev>
+            struct Empty<
+                queue::QueueGenericNonBlocking<TDev>>
+            {
+                //-----------------------------------------------------------------------------
+                ALPAKA_FN_HOST static auto empty(
+                    queue::QueueGenericNonBlocking<TDev> const & queue)
+                -> bool
+                {
+                    return queue.m_spQueueImpl->m_workerThread.isIdle();
+                }
+            };
+        }
+    }
+}
+
+#include <alpaka/event/EventGeneric.hpp>
```

Comment:
This looks strange

Reply:
This is just copied from the implementation of QueueCpuNonBlocking. I guess someone wanted to have the queue header also provide the event definition. The Event header cannot included at the top though, because it requires the definition of the queue for the trait specializations. I did not try removing it, assuming it may break other code.

Reply:
I am fine with this